src/discof/repair/fd_repair_tile.c

```diff
@@ -355,12 +356,265 @@ repair_shred_deliver( fd_shred_t const *            shred,
+
+  /* Sign it */
+  repair_signer( repair_tile_ctx, pong->signature.uc, pre_image, FD_PING_PRE_IMAGE_SZ, FD_KEYGUARD_SIGN_TYPE_SHA256_ED25519 );
+
+  fd_bincode_encode_ctx_t ctx;
+  uchar buf[1024];
+  ctx.data = buf;
+  ctx.dataend = buf + sizeof(buf);
+  FD_TEST(0 == fd_repair_protocol_encode(&protocol, &ctx));
+  ulong buflen = (ulong)((uchar*)ctx.data - buf);
+
+  repair_send_intake_packet( buf, buflen, peer_addr, self_ip4_addr, repair_tile_ctx );
+}
+
+/* Pass a raw client response packet into the protocol. addr is the address of the sender */
+static int
+fd_repair_recv_clnt_packet( fd_repair_tile_ctx_t * repair_tile_ctx,
+                           fd_repair_t *                 glob,
+                            uchar const *                 msg,
+                            ulong                         msglen,
+                            fd_repair_peer_addr_t const * src_addr,
+                            uint                          dst_ip4_addr ) {
+  glob->metrics.recv_clnt_pkt++;
+
+  FD_SCRATCH_SCOPE_BEGIN {
+    while( 1 ) {
+      fd_bincode_decode_ctx_t ctx = {
+        .data    = msg,
+        .dataend = msg + msglen
+      };
+
+      ulong total_sz = 0UL;
+      if( FD_UNLIKELY( fd_repair_response_decode_footprint( &ctx, &total_sz ) ) ) {
+        /* Solana falls back to assuming we got a shred in this case
+           https://github.com/solana-labs/solana/blob/master/core/src/repair/serve_repair.rs#L1198 */
+        break;
+      }
+
+      uchar * mem = fd_scratch_alloc( fd_repair_response_align(), total_sz );
+      if( FD_UNLIKELY( !mem ) ) {
+        FD_LOG_ERR(( "Unable to allocate memory for repair response" ));
+      }
+
+      fd_repair_response_t * gmsg = fd_repair_response_decode( mem, &ctx );
+      if( FD_UNLIKELY( ctx.data != ctx.dataend ) ) {
+        break;
+      }
+
+      switch( gmsg->discriminant ) {
+      case fd_repair_response_enum_ping:
+        fd_repair_handle_ping( repair_tile_ctx, glob, &gmsg->inner.ping, src_addr, dst_ip4_addr );
+        break;
+      }
+
+
+      return 0;
+    }
+
+    // are we still using this??
+    /* Look at the nonse */
+    if( msglen < sizeof(fd_repair_nonce_t) ) {
+      return 0;
+    }
+    ulong shredlen = msglen - sizeof(fd_repair_nonce_t); /* Nonce is at the end */
+    fd_repair_nonce_t key = *(fd_repair_nonce_t const *)(msg + shredlen);
+    fd_needed_elem_t * val = fd_needed_table_query( glob->needed, &key, NULL );
+    if( NULL == val ) {
+
+      return 0;
+    }
+
+    fd_active_elem_t * active = fd_active_table_query( glob->actives, &val->id, NULL );
+    if( NULL != active ) {
+      /* Update statistics */
+      active->avg_reps++;
+      active->avg_lat += glob->now - val->when;
+    }
+
+    fd_shred_t const * shred = fd_shred_parse(msg, shredlen);
+
+    if( shred == NULL ) {
+      FD_LOG_WARNING(("invalid shread"));
+    } else {
+      repair_shred_deliver(shred, shredlen, src_addr, &val->id, repair_tile_ctx);
```

Comment:
similar to above, i think a cleaner API would be to not actually call `repair_shred_deliver`

(also we can probably clean this up. the mutation of active table should be managed by caller separately vs. inside this function. but this can happen in a later PR)